src/ws/session.c

```diff
@@ -716,15 +726,18 @@ main (int argc,
         [1] = 2,  /* write stdout to stderr (avoid noise in ~/.profile) */
         [2] = -1, /* leave stderr alone */
         [3] = 1,  /* our stdout becomes cockpit-bridge fd3 */
+        [4] = fileno (login_messages) /* the memfd */
```

Comment:
I see, so this is where there is another remapping. But even without the previous patch this would still only be one extra dup2() call, instead of this whole "two dynamic loops" logic?

Reply:
It's not so easy: in the event of the `login_messages` file descriptor being `4` then `dup2(4,4)` is a no-op, which means that if `O_CLOEXEC` were to be set on that file descriptor, it wouldn't be cleared.  That's not currently the case (the `memfd_create()` call in `session-utils.c` doesn't set it) but depending on that to be true is extremely brittle.

We got bit by this "it's just a simple case of a `dup()` call or two..." once before.  I like the full-strength independently-reasoned-about universally-correct approach for exactly this reason.